mitxgraders/formulagrader/formulagrader.py

```diff
@@ -696,34 +744,104 @@ def scoped_eval(expression,
             # Compute expressions
             comparer_params_eval = self.eval_and_validate_comparer_params(
                 scoped_eval, comparer_params, siblings_eval)
+            comparer_params_evals.append(comparer_params_eval)
 
             # Before performing student evaluation, scrub the siblings
             # so that students can't use them
             for key in siblings_eval:
                 del varlist[key]
 
-            student_eval, used = scoped_eval(student_input)
+            student_eval, meta = scoped_eval(student_input)
+            student_evals.append(student_eval)
 
-            # Check if expressions agree
-            comparer_result = comparer(comparer_params_eval, student_eval, self.comparer_utils)
-            comparer_result = ItemGrader.standardize_cfn_return(comparer_result)
             if self.config['debug']:
                 # Put the siblings back in for the debug output
                 varlist.update(siblings_eval)
-                self.log_sample_info(i, varlist, funclist, student_eval,
-                                     comparer, comparer_params_eval, comparer_result)
+                self.log_eval_info(i, varlist, funclist, comparer_params_eval, student_eval)
 
-            if not comparer_result['ok']:
+        return comparer_params_evals, student_evals, meta.functions_used
+
+    def compare_evaluations(self, compare_parms_evals, student_evals, comparer, utils):
+        """
+        Compare the student evaluations to the expected results.
+        """
+        results = []
+        if isinstance(comparer, CorrelatedComparer):
+            result = comparer(compare_parms_evals, student_evals, utils)
+            results.append(ItemGrader.standardize_cfn_return(result))
+        else:
+            for compare_parms_eval, student_eval in zip(compare_parms_evals, student_evals):
+                result = comparer(compare_parms_eval, student_eval, utils)
+                results.append(ItemGrader.standardize_cfn_return(result))
+
+
+        if self.config['debug']:
+            self.log_comparison_info(comparer, results)
+
+        return results
+
+    @staticmethod
+    def consolidate_results(results, answer, failable_evals):
+        """
+        Consolidate comparer result(s) into just one result.
+
+        Arguments:
+            results: a list of results dicts
+            answer (dict): correctness data for the expected answer
+            failable_evals: int
+        """
+
+        # answer contains extra keys, so prune them
+        pruned_answer = { key: answer[key] for key in ['ok', 'grade_decimal', 'msg'] }
```

Comment:
Whenever `pruned_answer` is returned, it ignores any `msg` entries that the comparer may have generated. This is probably ok, but we should document that `msg` entries from comparers are ignored if the comparer finds agreement. What do you think?

Reply:
Yes, let's document it. I do think this is the correct behavior...I was tempted to concatenate the two messages, but just because a comparer returns `'ok': True` doesn't mean the input is correct (i.e., comparing could be checking one of several `answers=...` partial credit values).

Really, all error messages from comparers suffer this problem, hence my terrible "The submitted answer differs from an expected answer by a constant factor.".

Reply:
Perhaps this belongs in the general discussion regarding comparers then. It seems to be an appropriate policy.

Reply:
(I do realize that we aren't writing docs in this PR, but perhaps add it to a docstring?)